merlin-server/sql/merlin/plan_merge.sql

```diff
@@ -0,0 +1,592 @@
+  language plpgsql as $$
+  begin
+    if(select id from merge_request where id = request_id) is null then
+      raise exception 'Invalid merge request id %.', request_id;
+    end if;
+
+    if (select status from merge_request where id = request_id) != 'in-progress' then
+      raise exception 'Cannot reject merge not in progress.';
+    end if;
+
+    delete from conflicting_activities where merge_request_id = request_id;
+    delete from merge_staging_area where merge_staging_area.request_id = deny_merge.request_id;
+
+    update merge_request
+    set status = 'rejected'
+    where merge_request.id = request_id;
+
+    update plan
+    set is_locked = false
+    where plan.id = (select plan_id_receiving_changes from merge_request where id = request_id);
+  end
+  $$;
+
+/*
+  - Discard everything that was in the staging area
+  - Then, unlock the to-be-edited plan
+  - Then, change the merge request's status to 'pending'
+*/
+create or replace procedure cancel_merge(request_id integer)
+  language plpgsql as $$
+  declare
+    verify_status merge_request_status;
+begin
+    if(select id from merge_request where id = request_id) is null then
+      raise exception 'Invalid merge request id %.', request_id;
+    end if;
+
+    select status from merge_request where id = request_id into verify_status;
+    if not (verify_status = 'in-progress' or verify_status = 'pending') then
+      raise exception 'Cannot cancel merge.';
+    end if;
+
+    delete from conflicting_activities where merge_request_id = request_id;
+    delete from merge_staging_area where merge_staging_area.request_id = cancel_merge.request_id;
+
+    update merge_request
+    set status = 'pending'
+    where merge_request.id = request_id;
+
+    update plan
+    set is_locked = false
+    where plan.id = (select plan_id_receiving_changes from merge_request where id = request_id);
+end
+$$;
+
+create view merge_review_activity as
+select merge_request_id,
+       activity_id,
+       change_type_supplying as change_type_source,
+       change_type_receiving as change_type_target,
+       case resolution
+         when 'supplying'
+           then 'source'
+         when 'receiving'
+           then 'target'
+         end as resolution,
+       snap_act as source,
+       act as target,
+       'true'::boolean as is_conflict
+from conflicting_activities c
+       join merge_request m on c.merge_request_id = m.id
+       left join plan_snapshot_activities snap_act on m.snapshot_id_supplying_changes = snap_act.snapshot_id and c.activity_id = snap_act.id
+       left join activity_directive act on m.plan_id_receiving_changes = act.plan_id and c.activity_id = act.id
+union all
+select request_id as merge_request_id,
+       activity_id,
+       change_type as change_type_source,
+       'none' as change_type_target,
+       'source' as resolution,
+       snap_act as source,
+       act as target,
+       'false'::boolean as is_conflict
```

Comment:
Can someone explain this "is_conflict = false" case to me a bit more? Trying to understand if this needs to be displayed in the client.

Reply:
Or maybe the join to create this view here has an issue with the logic?

Reply:
That's a fair point. The thought process there was that non-conflicting activities have an implicit resolution of `source`, so that after a merge is committed, the mental model is that all of the activities with a resolution of `source` will match the `source` plan, and all activities with a resolution of `target` will match the version in the `target` plan.

This sort of makes sense thinking forward to a future capability to selectively accept/reject all changes (not just conflicts), but I can see that it doesn't make a lot of sense now.

Happy to rework this to better match UI needs/make invalid states unrepresentable. This view could be simplified to only roll up the conflicting activities, and then the UI would make two queries (which it could do with only one round trip) - one to get the non-conflicting activities from `merge_staging_area`, and one to get conflicting activities.